File: app/src/main/java/com/globalwarming/earthsaver/directories/DirectoryActivity.java
package com.globalwarming.earthsaver.directories;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.globalwarming.earthsaver.R;
import com.globalwarming.earthsaver.Util;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.WriteBatch;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DirectoryActivity extends AppCompatActivity {

    private ConstraintLayout parent;
    private Toolbar toolbar;
    private RecyclerView recyclerView;
    private ProgressDialog progressDialog;

    private DirectoryQuestionAdapter adapter;

    private FirebaseAuth mAuth;
    private FirebaseFirestore db;

    private List<DirectoryQuestion> list = new ArrayList<>();
    private List<DirectoryQuestion> oldRespondedQuestions = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_directory);

        String category = getIntent().getStringExtra("category");

        parent = findViewById(R.id.parent);
        toolbar = findViewById(R.id.toolbar);
        recyclerView = findViewById(R.id.recycler_view);
        setSupportActionBar(toolbar);
        toolbar.setTitle(category);
        toolbar.setNavigationOnClickListener(v -> {
            onBackPressed();
        });
        recyclerView.setHasFixedSize(true);
        adapter = new DirectoryQuestionAdapter();
        recyclerView.setAdapter(adapter);

        mAuth = FirebaseAuth.getInstance();
        db = FirebaseFirestore.getInstance();

        progressDialog = new ProgressDialog(this);
        progressDialog.setCancelable(false);
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.setMessage("Fetching questions ...");
        progressDialog.show();

        //fetchAllPrevious();
        db.collection("directories")
                .whereEqualTo("category", category)
                .get()
                .addOnCompleteListener(task -> {
                    progressDialog.dismiss();
                    if (task.isSuccessful()) {
                        for (QueryDocumentSnapshot document : task.getResult()) {
                            DirectoryQuestion directoryQuestion = document.toObject(DirectoryQuestion.class);
                            directoryQuestion.setId(document.getId());
                            list.add(directoryQuestion);
                        }
                        adapter.setList(list);
                    }
                });

    }

    private void fetchAllPrevious() {
        Long minTimestamp = System.currentTimeMillis() - (60 * 60 * 24 * 1000 * 30);
        db.collection("users")
                .document(mAuth.getCurrentUser().getUid())
                .collection("entries")
                .whereGreaterThan("timestamp", minTimestamp)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        for (QueryDocumentSnapshot document : task.getResult()) {
                            DirectoryQuestion directoryQuestion = document.toObject(DirectoryQuestion.class);
                            oldRespondedQuestions.add(directoryQuestion);
                        }
                    }


                });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_directory, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.menu_submit) {
            Util.hideKeyboard(parent);
            progressDialog.setMessage("Submitting answers ...");
            progressDialog.show();

            int points = 0;
            HashMap<String, Pair<Boolean, String>> entries = adapter.getEntries();
            String userId = mAuth.getUid();
            WriteBatch batch = db.batch();
            for (Map.Entry<String, Pair<Boolean, String>> entry : entries.entrySet()) {
                if (entry.getValue().first) {
                    points++;
                }
                DirectoryEntry directoryEntry = new DirectoryEntry();
                directoryEntry.setId(entry.getKey());
                directoryEntry.setTimestamp(System.currentTimeMillis());
                directoryEntry.setAnswer(entry.getValue().first);
                if (entry.getValue().first) {
                    directoryEntry.setPoint(1L);
                    directoryEntry.setNote(entry.getValue().second);
                } else {
                    directoryEntry.setPoint(0L);
                }
                batch.set(db.collection("users")
                        .document(userId)
                        .collection("entries")
                        .document(entry.getKey() + "#" + directoryEntry.getTimestamp()), directoryEntry);
            }
            batch.update(db.collection("users").document(mAuth.getUid()), "points", FieldValue.increment(points));

            batch.commit().addOnCompleteListener(task -> {
                progressDialog.dismiss();
                if (task.isSuccessful()) {
                    Toast.makeText(DirectoryActivity.this, "Successfully uploaded", Toast.LENGTH_SHORT).show();
                    finish();
                } else {
                    Snackbar.make(parent, "Some error occurred", Snackbar.LENGTH_SHORT).show();
                }
            });
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        if (!adapter.getEntries().isEmpty()) {
            AlertDialog alertDialog = new AlertDialog.Builder(this)
                    .setTitle("Warning")
                    .setMessage("Are you sure you want to go back? Your answers won't be submitted if you go back!")
                    .setNegativeButton("Yes", (dialogInterface, i) -> {
                        dialogInterface.dismiss();
                        super.onBackPressed();
                    })
                    .setPositiveButton("No", (dialogInterface, i) -> {
                        dialogInterface.dismiss();
                    })
                    .create();
            alertDialog.show();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onDestroy() {
        if (progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
        super.onDestroy();
    }

}

File: app/src/main/java/com/globalwarming/earthsaver/directories/DirectoryActivity.kt
package com.globalwarming.earthsaver.directories

import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.globalwarming.earthsaver.*
import com.globalwarming.earthsaver.R
import com.globalwarming.earthsaver.databinding.ActivityDirectoryBinding
import com.globalwarming.earthsaver.util.Loader
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.*
import java.util.*

class DirectoryActivity : AppCompatActivity() {

    private lateinit var binding: ActivityDirectoryBinding
    private lateinit var loader: Loader
    private lateinit var adapter: DirectoryQuestionAdapter
    private lateinit var mAuth: FirebaseAuth
    private lateinit var db: FirebaseFirestore
    private val list = mutableListOf<DirectoryQuestion>()
    private val oldRespondedQuestions = mutableListOf<DirectoryQuestion>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_directory)
        setSupportActionBar(binding.toolbar)

        val category = intent.getStringExtra("category")

        binding.toolbar.title = category
        binding.toolbar.setNavigationOnClickListener {
            onBackPressed()
        }
        binding.recyclerView.setHasFixedSize(true)
        adapter = DirectoryQuestionAdapter()
        binding.recyclerView.adapter = adapter
        mAuth = FirebaseAuth.getInstance()
        db = FirebaseFirestore.getInstance()
        loader = Loader(this, "Fetching questions ...")
        loader.show()

        fetchAllPrevious {
            db.collection("directories")
                .whereEqualTo("category", category)
                .get()
                .addOnCompleteListener { task ->
                    loader.dismiss()
                    if (task.isSuccessful) {
                        for (document in task.result!!) {
                            val directoryQuestion = document.toObject(DirectoryQuestion::class.java)
                            directoryQuestion.id = document.id
                            val contains = oldRespondedQuestions.firstOrNull {
                                it.id.split("#")[0] == document.id
                            } != null
                            directoryQuestion.canRespond = !contains
                            list.add(directoryQuestion)
                        }
                        adapter.setList(list.sortedByDescending { it.canRespond })
                    }
                }
        }

    }

    private fun fetchAllPrevious(cb: () -> Unit) {
        val minTimestamp = System.currentTimeMillis() - (24 * 60 * 60 * 1000)
        db.collection("users")
            .document(mAuth.currentUser!!.uid)
            .collection("entries")
            .whereGreaterThan("timestamp", minTimestamp)
            .get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result!!) {
                        val directoryQuestion = document.toObject(DirectoryQuestion::class.java)
                        directoryQuestion.id = document.id
                        oldRespondedQuestions.add(directoryQuestion)
                    }
                    cb()
                } else {
                    cb()
                }
            }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_directory, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_submit) {
            Util.hideKeyboard(binding.parent)
            loader.setText("Submitting answers ...")
            loader.show()
            var points = 0L
            val entries = adapter.getEntries()
            val userId = mAuth.uid!!
            val batch = db.batch()
            for ((key, value) in entries) {
                val dq = list.firstOrNull {
                    it.id == key
                }
                if (dq != null) {
                    points += dq.points
                    val directoryEntry = DirectoryEntry()
                    directoryEntry.id = key
                    directoryEntry.timestamp = System.currentTimeMillis()
                    directoryEntry.answer = true
                    directoryEntry.point = dq.points
                    directoryEntry.note = value
                    batch[db.collection("users")
                        .document(userId)
                        .collection("entries")
                        .document(key + "#" + directoryEntry.timestamp)] = directoryEntry
                }
            }
            batch.update(
                db.collection("users").document(mAuth.uid!!),
                "points",
                FieldValue.increment(points)
            )
            batch.commit().addOnCompleteListener { task ->
                loader.dismiss()
                if (task.isSuccessful) {
                    Toast.makeText(
                        this@DirectoryActivity,
                        "Successfully uploaded",
                        Toast.LENGTH_SHORT
                    ).show()
                    finish()
                } else {
                    Snackbar.make(binding.parent, "Some error occurred", Snackbar.LENGTH_SHORT).show()
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (adapter.getEntries().isNotEmpty()) {
            val alertDialog = AlertDialog.Builder(this)
                .setTitle("Warning")
                .setMessage("Are you sure you want to go back? Your answers won't be submitted if you go back!")
                .setNegativeButton("Yes") { dialog, _ ->
                    dialog.dismiss()
                    super.onBackPressed()
                }
                .setPositiveButton("No") { dialog, _ -> dialog.dismiss() }
                .create()
            alertDialog.show()
        } else {
            super.onBackPressed()
        }
    }

    override fun onDestroy() {
        if (loader.isShowing) {
            loader.dismiss()
        }
        super.onDestroy()
    }
}